### PR TITLE
Proposal War support with Gradle

### PR DESCRIPTION
**Pull request by CyrilleH**
Towards #431, here's my proposal to support "WAR" project with Jib.

~~If the "webAppRoot" is filled, these layers are relocated as following :~~
~~- "resources"  --> _webAppRoot_/WEB-INF/classes~~
~~- "classes"  --> _webAppRoot_/WEB-INF/classes~~
~~- "dependencies" --> _webAppRoot_/WEB-INF/lib~~
~~- "snapshot dependencies" --> _webAppRoot_/WEB-INF/lib~~

~~I've created 3 more layers :~~
~~- "webapp" for JSP files, assets, angular generated files, ...~~
~~- "meta-inf" for files like "context.xml"~~
~~- "web-inf" for files like "web.xml", etc...~~

~~I'm going to rebase tonight on master to replace my implementation of "entrypoint" with yours.~~

Upon validation, I will consolidate with more tests, clean-ups and maybe an implementation for the maven plugin.

I've performed some real life tests with 10 different  "Tomcat" projects.~~

### CODE REVIEW

**Comment by chanseokoh** [comment]
Thanks for your contribution!

We've done and reviewed an internal design for WAR support, and our design has one main difference from yours: we decided that we will use and rely on WAR assembly plugins to package a WAR instead of packaging ourselves from the webapp source directory, compiled `.class` files, etc. Below are some rationales for the decision.

We thought that it is impractical to extend Jib to become another complicated WAR-packaging plugin. We also believe conforming to the de-facto WAR-packaging plugins provides more complete support for real-world WAR-use cases. Note some examples of what the `maven-war-plugin` does:
- Adding arbitrary external web resources with `<webResources>`
   - Can configure where to pull (`<directory>`) and place (`<targetPath>`) resources
   - Can also configure `<includes>`/`<excludes>` and `<filtering>` (interpolating property values in some files).
- `web.xml` filtering with `<filteringDeploymentDescriptors>` (interpolating properties values)
- Sharing common resources across multiple web applications (overlays)
- Including and excluding files from the final WAR with `<packagingIncludes>` and `<packagingExcludes>`
- Placing dependency JARs under `WEB-INF/lib` depending on `<scope>` or `<optional>` values of `<dependency>`s in `pom.xml`.

That said, could you revise your PR to use and take the final WAR produced by the Gradle WAR Plugin? We've recently checked in the `ZipUtil` class into your codebase, which you can use to unpack a WAR. You'll be able to see `ZipUtil` once your merge with the HEAD.

**Comment by CyrilleH** [comment]
I signed it!

Le lun. 3 sept. 2018 à 15:59, Chanseok Oh <notifications@github.com> a
écrit :

**Comment by CyrilleH** [comment]
If I understood you correctly (sorry I'm French ), you want a design like this : 
1) let gradle generate the war file (with the standard plugin)
2) unzip it in a temporary folder (with com.google.cloud.tools.jib.plugins.common.ZipUtil)
3) and then, create the "dependencies", "classes", "snapshot dependencies", "webapp" layers

It's an another solution but it creates a lot of duplicate/temporary files. The Gradle plugins avoid generating this usually.

Nevertheless I agree with you, it would be better to rely on the information given in the war definition.

**Comment by chanseokoh** [comment]
> If I understood you correctly (sorry I'm French ), you want a design like this:

Yes, exactly. It would have been nice if there were a way to avoid unpacking the WAR, but I couldn't really find such a way so far. (Not that I really looked into Gradle thoroughly for this.)

> Also, I think it's better to let the user define the webapp root in the container.

Our design did not allow this, at least for the very first implementation (we were exploring if there could be a way of zero-configuration), but I guess allowing to configure the webapp location in an image may actually be useful to quickly support arbitrary base images. We will discuss this approach internally.

**Comment by chanseokoh** [comment]
> 3.and then, create the "dependencies", "classes", "snapshot dependencies", "webapp" layers

Oh, forgot to mention that we think
- `WEB-INF/classes/***/*.class` go into the "classes" layer (probably excluding non-`.class` files)
- `WEB-INF/lib/*-SNAPSHOT.jar` into "snapshot dependencies"
- Other `WEB-INF/lib/*.jar` into "dependencies"
- All other files into "resources"

I'm wondering if you also think this makes sense.

**Comment by chanseokoh** [comment]
@CyrilleH just checking in, as I haven't heard back from you for a while. Are you working on this?

And just a heads-up: we had an internal discussion about the `webAppRoot` configuration parameter, and we decided we won't add it, at least in this PR.

**Comment by CyrilleH** [comment]
I thought you rejected my proposal.
Since it works very well for us, I decided to wait a little bit to understand the direction you take.
Do you want to support only Jetty (for example) ? Do you want to implement only one webapp layer ?

**Comment by CyrilleH** [comment]
Ok, I'll try to implement something this weekend.

**Comment by loosebazooka** [comment]
exploding a war shouldn't be too hard: https://github.com/GoogleCloudPlatform/app-gradle-plugin/blob/master/src/main/java/com/google/cloud/tools/gradle/appengine/standard/ExplodeWarTask.java

**Comment by CyrilleH** [comment]
Yes, "Sync Task" is the best way to explode a War with Gradle.

If I understand this (issue)[https://github.com/GoogleContainerTools/jib/issues/964] correctly, I have to change my "jib.container.webAppRoot" parameter into "jib.container.appRoot".

**Comment by chanseokoh** [comment]
> If I understand this [issue](https://github.com/GoogleContainerTools/jib/issues/964) correctly, I have to change my "jib.container.webAppRoot" parameter into "jib.container.appRoot".

I am taking care of #964 with #984. You might want to merge with that branch.

**Comment by CyrilleH** [comment]
Ok thanks.

Maybe I could use something like that (and work directly with the war file) : 
`
import java.nio.file.FileSystem;
import java.nio.file.FileSystems;

final FileSystem zipfs = FileSystems.newFileSystem(war.getArchivePath().toPath(), JavaLayerConfigurations.class.getClassLoader());
Path dependenciesFiles = zipfs.getPath("/WEB-INF/lib");
try (Stream<Path> stream = Files.list(dependenciesFiles)) {
        stream.forEach(dependenciesFiles::add);
}
`

I'm going to do some tests to see if it's a good solution.

**Comment by CyrilleH** [comment]
I did a second implementation (with WAR unzipping).
@chanseokoh, I rebased my commit on your "i964-container.appRoot" branch.

Please let me know if it's good enough for you.

When your branch will be merged on master, I will rebase my branch.

**Comment by chanseokoh** [comment]
@CyrilleH thanks for your time for implementing this. Haven't looked into the code yet, but I think we're going to the right direction. Once #964 is merged into master, we will be able to start reviewing this PR in detail. So I will have to ask you to merge with master later once #964 is done. I'll try to get #964 in ASAP.

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java [hunk 1]
```diff
@@ -58,45 +63,86 @@ static JavaLayerConfigurations getForProject(
...
+    final boolean isWebapp = GradleProjectProperties.getWarTask(project) != null;
 
-    // Adds each file in each classes output directory to the classes files list.
-    FileCollection classesOutputDirectories = mainSourceSet.getOutput().getClassesDirs();
-    gradleJibLogger.info("Adding corresponding output directories of source sets to image");
-    for (File classesOutputDirectory : classesOutputDirectories) {
-      if (Files.notExists(classesOutputDirectory.toPath())) {
-        gradleJibLogger.info("\t'" + classesOutputDirectory + "' (not found, skipped)");
-        continue;
+    if (isWebapp) {
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java]: This method became too long, so let's break it down. Some notes:

- No `final` on local variables (`isWebapp`).
- `isWebapp` --> `isWebApp`. Perhaps `isWarProject`.

```java
  static JavaLayerConfigurations getForProject(
      Project project, GradleJibLogger logger, Path extraDirectory, String appRoot)
      throws IOException {
    boolean isWarProject = GradleProjectProperties.getWarTask(project) != null;
    if (isWarProject) {
      logger.info("WAR project identified, creating WAR image: " + project.getDisplayName());
      return getForWarProject(war, logger, extraDirectory, appRoot);
    } else {
      return getForJarProject(project, logger, extraDirectory, appRoot);
    }
  }
```
where `getForJarProject()` is the previous `getForProject()` (I don't think you will touch anything on it).

File: jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibPluginTest.java [hunk 2]
```diff
@@ -148,4 +150,77 @@ public void testProjectDependencyAssembleTasksAreRun() {
...
+            .iterator()
+            .next());
+    Assert.assertEquals(
+        JibPlugin.DEFAULT_WEBAPP_FROM_IMAGE,
+        ((BuildImageTask) rootProject.getTasks().getByPath(JibPlugin.BUILD_IMAGE_TASK_NAME))
+            .getJib()
+            .getBaseImage());
+  }
+
+  @Test(expected = UnknownTaskException.class)
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibPluginTest.java]: We don't use `expected` in this codebase. Please follow the practice below:

```java
try {
   theVeryMethodThatThrowsAnException(); // should be narrowly scoped.
   Assert.fail();
} catch (UnknownTaskException ex) {
   Assert.assertNotNull(ex.getMessage);
}
```

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/PluginConfigurationProcessor.java [hunk 3]
```diff
@@ -124,6 +135,32 @@ static PluginConfigurationProcessor processCommonConfiguration(
...
 
+  /**
+   * Get the user specified entrypoint or the default one
+   *
+   * @param logger the logger used to display messages.
+   * @param jibExtension the {@link JibExtension} providing the configuration data
+   * @param projectProperties used for providing additional information
+   * @return the entrypoint
+   */
+  static List<String> getEntrypoint(
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/PluginConfigurationProcessor.java]: I think `computeEntrypoint` fits better.

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/PluginConfigurationProcessor.java [hunk 4]
```diff
@@ -124,6 +135,32 @@ static PluginConfigurationProcessor processCommonConfiguration(
         fromCredential);
   }
 
+  /**
+   * Get the user specified entrypoint or the default one
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/PluginConfigurationProcessor.java]: The description isn't really correct, because the code may try to infer a main class. (For example, it may read `MANIFEST.MF` in a JAR.)

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java [hunk 5]
```diff
@@ -58,45 +63,86 @@ static JavaLayerConfigurations getForProject(
...
+      }
+      // Some files in WEB-INF need to be in resources directory (e.g. web.xml, ...)
+      Path webinfOutputDirectory = explodedWarPath.resolve(JavaLayerConfigurations.WEB_INF);
+      try (Stream<Path> fileStream = Files.list(webinfOutputDirectory)) {
+        fileStream
+            .filter(path -> !path.endsWith("classes") && !path.endsWith("lib"))
+            .forEach(resourcesFiles::add);
+      }
+    } else {
+      if (appRoot.isEmpty()) {
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java]: Can this happen? I thought `appRoot` is `DEFAULT_APP_ROOT` when not specified. Can you verify?

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java [hunk 6]
```diff
@@ -58,45 +63,86 @@ static JavaLayerConfigurations getForProject(
...
+    if (isWebapp) {
+      if (appRoot.isEmpty()) {
+        appRoot = ContainerConfiguration.DEFAULT_WEBAPP_ROOT;
       }
-      gradleJibLogger.info("\t'" + classesOutputDirectory + "'");
-      try (Stream<Path> classFileStream = Files.list(classesOutputDirectory.toPath())) {
+      Path explodedWarPath = GradleProjectProperties.getExplodedWarDirectory(project).toPath();
+      Path classesOutputDirectory =
+          explodedWarPath.resolve(JavaLayerConfigurations.WEB_INF_CLASSES);
+      try (Stream<Path> classFileStream = Files.list(classesOutputDirectory)) {
```
[the remark this answers is not in the document]
Comment (chanseokoh) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java]: However, if you feel you don't want to tackle this relatively minor problem in this PR, just file a new issue. We can fix this issue in a later PR, which may actually be better to make this PR short and help get the PR merged.

File: jib-gradle-plugin/README.md [hunk 7]
```diff
@@ -238,6 +238,31 @@ jib {
 }
 ```
 
+### Webapp Example with Apache Tomcat
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-gradle-plugin/README.md]: Ok, no problem.

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/ExplodedWarTask.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+   *
+   * @param explodedWarDirectory the directory where to extract the war file
+   */
+  public void setExplodedWarDirectory(File explodedWarDirectory) {
+    this.explodedWarDirectory = explodedWarDirectory;
+    into(explodedWarDirectory);
+  }
+
+  @OutputDirectory
+  public @Nullable File getExplodedWarDirectory() {
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/ExplodedWarTask.java]: Done

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/ExplodedWarTask.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
+package com.google.cloud.tools.jib.gradle;
+
+import java.io.File;
+import javax.annotation.Nullable;
+import org.gradle.api.tasks.OutputDirectory;
+import org.gradle.api.tasks.Sync;
+
+public class ExplodedWarTask extends Sync {
+  private @Nullable File explodedWarDirectory;
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/ExplodedWarTask.java]: Done

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaDockerContextGenerator.java [hunk 10]
```diff
@@ -139,42 +136,49 @@ private static String mapToDockerfileString(Map<String, String> map, String comm
...
    * @param javaLayerConfigurations the {@link JavaLayerConfigurations}
    */
   public JavaDockerContextGenerator(JavaLayerConfigurations javaLayerConfigurations) {
+    String appRoot = javaLayerConfigurations.getAppRoot();
+    appRoot = appRoot.endsWith("/") ? appRoot : appRoot + '/';
+
     ImmutableList.Builder<CopyDirective> copyDirectivesBuilder = ImmutableList.builder();
+    isWebapp = javaLayerConfigurations.isWebapp();
+    String dependenciesSuffix =
+        isWebapp
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaDockerContextGenerator.java]: Done

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java [hunk 11]
```diff
@@ -32,51 +33,86 @@
...
 public class JavaLayerConfigurations {
+  public static final String WEB_INF = "WEB-INF";
+  public static final String WEB_INF_LIB = WEB_INF + "/lib/";
+  public static final String WEB_INF_CLASSES = WEB_INF + "/classes/";
 
   /** Represents the different types of layers for a Java application. */
   @VisibleForTesting
   enum LayerType {
-    DEPENDENCIES("dependencies", JavaEntrypointConstructor.DEFAULT_DEPENDENCIES_PATH_ON_IMAGE),
+    DEPENDENCIES(
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java]: Done

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java (hunk 5 above)
Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java]: There is 2 default values : DEFAULT_WEBAPP_ROOT and DEFAULT_APP_ROOT.

So, I changed this code 

```
ContainerParameters {
    private String appRoot = DEFAULT_APP_ROOT;
}
```
with that : 

```
ContainerParameters {
    private String appRoot = "";
}
```

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/PluginConfigurationProcessor.java (hunk 4 above)
Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/PluginConfigurationProcessor.java]: Done

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/PluginConfigurationProcessor.java (hunk 3 above)
Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/PluginConfigurationProcessor.java]: Done

File: jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibPluginTest.java (hunk 2 above)
Comment (CyrilleH) [reply jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/JibPluginTest.java]: Done

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java (hunk 1 above)
Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java]: Done

**Comment by CyrilleH** [comment]
@chanseokoh I've rebased on your branch.

**Comment by chanseokoh** [comment]
@CyrilleH thanks! I just noticed that I need to further change `JavaLayerConfigurations` to allow configuring layers with more flexibility (#1007). I'll work on #1007 on top of #984.

File: jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java [hunk 12]
```diff
@@ -32,112 +32,132 @@
...
             LayerConfiguration.builder().setName(layerType.getName());
 
         // Adds all the layer files recursively.
         List<Path> layerFiles = Preconditions.checkNotNull(layerFilesMap.get(layerType));
         for (Path layerFile : layerFiles) {
-          layerConfigurationBuilder.addEntryRecursive(
-              layerFile, layerType.getExtractionPath().resolve(layerFile.getFileName()));
+          final Path pathInContainer;
+          Path subPath = Paths.get(extractionPath);
+          if (layerType.equals(LayerType.RESOURCES)
```
Comment (chanseokoh) [inline jib-core/src/main/java/com/google/cloud/tools/jib/frontend/JavaLayerConfigurations.java]: Urgh... I guess #984 isn't enough. I'll fix #1007, and after that, I think this class will become truly generic and there won't be any special code to handle the WAR case.

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java (hunk 5 above)
Comment (chanseokoh) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java]: Got it. We'll worry about this (if we have to) later.

**Comment by CyrilleH** [comment]
Yes, my code is not ideal in `JavaLayerConfigurations` : 
```
          Path subPath = Paths.get(extractionPath);
          if (layerType.equals(LayerType.RESOURCES)
              && layerFile.getParent() != null
              && layerFile.getParent().endsWith(WEB_INF)) {
            pathInContainer = subPath.resolve(WEB_INF).resolve(layerFile.getFileName());
          } else {
            pathInContainer = subPath.resolve(layerFile.getFileName());
          }
```

**Comment by chanseokoh** [comment]
@CyrilleH  just an update. Sorry for the delay. Fixing #1007 needed some more thoughts, as it is about changing the API to be general.

**Comment by chanseokoh** [comment]
#1007 is fixed. Please merge with master. Hopefully you won't need to tinker with `JavaLayerConfigurations` and `JavaDockerContextGenerator` anymore.

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java]: Fixed

**Comment by chanseokoh** [comment]
Thanks. Looks like this is in a good shape for review. @GoogleContainerTools/java-tools falks, PTAL.

**Comment by chanseokoh** [comment]
There are test failures. You can run the tests locally with `./build.sh`.

**Comment by CyrilleH** [comment]
Yes, it is an empty directory which is not commit.

**Comment by chanseokoh** [comment]
> Yes, it is an empty directory which is not commit.

Yeah, AFAIK, it is not possible to commit an empty directory in Git. There's the trick to add `.gitignore`, but I guess it's not applicable to us to work as a test case. I think you'll need to revise the test case that doesn't deal with an empty directory, unfortunately.

**Comment by coollog** [comment]
We use a `.gitkeep` file to keep an empty directory, such as in the `jib-core/src/test/resources/directoryA`.

**Comment by CyrilleH** [comment]
What do you think about this code in setUp() :
image
The empty directory is created in the build folder so for me it's good enough.

**Comment by chanseokoh** [comment]
OK, seems we can create it that way. Just remove the `if (Files.notExist())`. Generally tests should be decisive with no ambiguity or uncertainty.

**Comment by CyrilleH** [comment]
But, it's mandatory. If we don't test the directory existence, we can't re-play the test without cleaning the project.

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/ExplodedWarTask.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
+package com.google.cloud.tools.jib.gradle;
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/ExplodedWarTask.java]: Sorry, Intellij masked this header.

File: jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurationsTest.java [hunk 14]
```diff
@@ -238,4 +245,144 @@ public void testGetForProject_nonDefaultAppRoot() throws IOException, URISyntaxE
...
+    ImmutableList<Path> expectedResourcesFiles =
+        ImmutableList.of(
+            webappDirectory.resolve("jib-exploded-war/META-INF"),
+            webappDirectory.resolve("jib-exploded-war/META-INF/context.xml"),
+            webappDirectory.resolve("jib-exploded-war/Test.jsp"),
+            webappDirectory.resolve("jib-exploded-war/WEB-INF/classes/package/test.properties"),
+            webappDirectory.resolve("jib-exploded-war/WEB-INF/web.xml"));
+    ImmutableList<Path> expectedClassesFiles =
+        ImmutableList.of(
+            webappDirectory.resolve("jib-exploded-war/WEB-INF/classes/HelloWorld.class"));
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurationsTest.java]: Good idea

**Comment by CyrilleH** [comment]
@coollog thanks for your code review.
I pushed a fix.

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java [hunk 15]
```diff
@@ -40,8 +42,24 @@
   /** Name of the `main` {@link SourceSet} to use as source files. */
   private static final String MAIN_SOURCE_SET_NAME = "main";
 
+  private static final String SNAPSHOT = "SNAPSHOT";
+  private static final String WEB_INF = "WEB-INF";
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java]: Done

File: jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurationsTest.java [hunk 16]
```diff
@@ -135,6 +136,13 @@ public void setUp() throws URISyntaxException {
     Mockito.when(mockMainSourceSetOutput.getClassesDirs()).thenReturn(classesFileCollection);
     Mockito.when(mockMainSourceSetOutput.getResourcesDir()).thenReturn(resourcesOutputDir.toFile());
     Mockito.when(mockMainSourceSet.getRuntimeClasspath()).thenReturn(runtimeFileCollection);
+    // We can't commit an empty directory in Git, so it's created if it does not exist
+    Path empty_dir =
+        Paths.get(Resources.getResource("webapp").toURI())
+            .resolve("jib-exploded-war/WEB-INF/classes/empty_dir");
+    if (Files.notExists(empty_dir)) {
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurationsTest.java]: Done

File: jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurationsTest.java [hunk 17]
```diff
@@ -135,6 +136,13 @@ public void setUp() throws URISyntaxException {
     Mockito.when(mockMainSourceSetOutput.getClassesDirs()).thenReturn(classesFileCollection);
     Mockito.when(mockMainSourceSetOutput.getResourcesDir()).thenReturn(resourcesOutputDir.toFile());
     Mockito.when(mockMainSourceSet.getRuntimeClasspath()).thenReturn(runtimeFileCollection);
+    // We can't commit an empty directory in Git, so it's created if it does not exist
+    Path empty_dir =
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-gradle-plugin/src/test/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurationsTest.java]: Done

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/JibPlugin.java [hunk 18]
```diff
@@ -109,13 +125,41 @@ public void apply(Project project) {
...
         projectAfterEvaluation -> {
           try {
-            // Has all tasks depend on the 'classes' task.
-            Task classesTask = projectAfterEvaluation.getTasks().getByPath("classes");
-            buildImageTask.dependsOn(classesTask);
-            dockerContextTask.dependsOn(classesTask);
-            buildDockerTask.dependsOn(classesTask);
-            buildTarTask.dependsOn(classesTask);
-            filesTask.dependsOn(classesTask);
+            War warTask = GradleProjectProperties.getWarTask(project);
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/JibPlugin.java]: TODO added.

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleProjectProperties.java [hunk 19]
```diff
@@ -91,6 +91,14 @@ static War getWarTask(Project project) {
     return (War) warPluginConvention.getProject().getTasks().findByName("war");
   }
 
+  static File getExplodedWarDirectory(Project project) {
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleProjectProperties.java]: Done

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java [hunk 20]
```diff
@@ -132,5 +150,109 @@ static JavaLayerConfigurations getForProject(
...
+                    layerBuilder.addResourceFile(
+                        path, classesExtractionPath.resolve(srcWebInfClasses.relativize(path)));
+                  }
+                }
+              } else {
+                layerBuilder.addResourceFile(
+                    path, classesExtractionPath.resolve(srcWebInfClasses.relativize(path)));
+              }
+            } catch (IOException e) {
+              logger.warn("IOException : '" + path + "' not added in class/resource layer", e);
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java]: Done

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java [hunk 21]
```diff
@@ -132,5 +150,109 @@ static JavaLayerConfigurations getForProject(
...
+    try (Stream<Path> fileStream = Files.walk(srcWebInfClasses)) {
+      fileStream.forEach(
+          path -> {
+            try {
+              if (FileSystems.getDefault().getPathMatcher("glob:**.class").matches(path)) {
+                layerBuilder.addClassFile(
+                    path, classesExtractionPath.resolve(srcWebInfClasses.relativize(path)));
+
+              } else if (Files.isDirectory(path)) {
+                try (DirectoryStream<Path> dirStream = Files.newDirectoryStream(path)) {
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java]: Done

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java [hunk 22]
```diff
@@ -132,5 +150,109 @@ static JavaLayerConfigurations getForProject(
...
+
+    Builder layerBuilder = JavaLayerConfigurations.builder();
+    AbsoluteUnixPath dependenciesExtractionPath = appRoot.resolve(WEB_INF_LIB);
+    AbsoluteUnixPath classesExtractionPath = appRoot.resolve(WEB_INF_CLASSES);
+    AbsoluteUnixPath webInfExtractionPath = appRoot.resolve(WEB_INF);
+
+    // For "WEB-INF/classes", *.class in class layer, other files and empty directories in resource
+    // layer
+    Path srcWebInfClasses = explodedWarPath.resolve(WEB_INF_CLASSES);
+    try (Stream<Path> fileStream = Files.walk(srcWebInfClasses)) {
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java]: Yes, it's better.

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java [hunk 23]
```diff
@@ -40,8 +42,24 @@
...
+      logger.info("WAR project identified, creating WAR image: " + project.getDisplayName());
+      return getForWarProject(project, logger, extraDirectory, appRoot);
+    } else {
+      return getForJarProject(project, logger, extraDirectory, appRoot);
+    }
+  }
+
   /**
-   * Resolves the source files configuration for a Gradle {@link Project}.
+   * Resolves the source files configuration for a Jar Gradle {@link Project}.
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java]: Done

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java [hunk 24]
```diff
@@ -40,8 +42,24 @@
   /** Name of the `main` {@link SourceSet} to use as source files. */
   private static final String MAIN_SOURCE_SET_NAME = "main";
 
+  private static final String SNAPSHOT = "SNAPSHOT";
+  private static final String WEB_INF = "WEB-INF";
+  private static final String WEB_INF_LIB = WEB_INF + "/lib/";
+  private static final String WEB_INF_CLASSES = WEB_INF + "/classes/";
+
+  static JavaLayerConfigurations getForProject(
```
[the remark this answers is not in the document]
Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/GradleLayerConfigurations.java]: Done

**Comment by chanseokoh** [comment]
Didn't look into every test detail, but this looks OK-ish. I'd like to update some Javadocs/comments, rename variables, refactor some code, but I think I can take care of those things after merging this.

@coollog if you are good with the new way that the default base image and appRoot is set (see https://github.com/GoogleContainerTools/jib/pull/923/files#diff-9266d6136413955a86ce0592f20891f6R44 and https://github.com/GoogleContainerTools/jib/pull/923/files#diff-aef8bf567c91df6622aa0ea29ed0c682R133), I think we can merge this soon?

**Comment by coollog** [comment]
Okay LGTM. We should add an integration test for a Gradle war project to `JibPluginIntegrationTest` but that can be in a follow-up PR.

**Comment by chanseokoh** [comment]
About adding integration tests, we are keeping track of it, so we're good: https://github.com/GoogleContainerTools/jib/issues/431#issuecomment-422827669

I'll merge this PR maybe on Monday. Anyone can review this until then. I will be working on some cleanups on my side on top of this PR.

File: jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/ContainerParameters.java [hunk 25]
```diff
@@ -42,7 +41,7 @@
   private ImageFormat format = ImageFormat.Docker;
   private List<String> ports = Collections.emptyList();
   private Map<String, String> labels = Collections.emptyMap();
-  private String appRoot = JavaLayerConfigurations.DEFAULT_APP_ROOT;
+  private String appRoot = "";
```
Comment (chanseokoh) [inline jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/ContainerParameters.java]: I'm thinking this may affect Maven. Gradle now sets this value [below](https://github.com/GoogleContainerTools/jib/pull/923/files#diff-aef8bf567c91df6622aa0ea29ed0c682R136), but I don't see Maven doing a similar setup in this PR. Need to confirm if Maven works without any problem.

Comment (CyrilleH) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/ContainerParameters.java]: I think there is no side effect.
In this inner class `com.google.cloud.tools.jib.maven.JibPluginConfiguration.ContainerParameters`, there is still this code : 
```
@Parameter private String appRoot = JavaLayerConfigurations.DEFAULT_APP_ROOT;
```

Comment (chanseokoh) [reply jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/ContainerParameters.java]: Oops, you're right. The file is not shared, but exclusive to the Gradle plugin.
